add x32 ABI support (x86_64 -mx32)

It's a forward of downstream https://bugs.gentoo.org/614374 report where users try to get `libunwind` built on x32 userland. libffi fails there as:

```
~/libunwind $ LANG=C make
Making all in src
make[1]: Entering directory '/root/libunwind/src'
make  all-am
make[2]: Entering directory '/root/libunwind/src'
depbase=`echo x86_64/Los-linux.lo | sed 's|[^/]*$|.deps/&|;s|\.lo$||'`;\
/bin/sh ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../include  -I../include -I../include/tdep-x86_64 -I. -D_GNU_SOURCE -DNDEBUG  -g -O2 -fexceptions -Wall -Wsign-compare -D__EXTENSIONS__ -MT x86_64/Los-linux.lo -MD -MP -MF $depbase.Tpo -c -o x86_64/Los-linux.lo x86_64/Los-linux.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../include -I../include -I../include/tdep-x86_64 -I. -D_GNU_SOURCE -DNDEBUG -g -O2 -fexceptions -Wall -Wsign-compare -D__EXTENSIONS__ -MT x86_64/Los-linux.lo -MD -MP -MF x86_64/.deps/Los-linux.Tpo -c x86_64/Los-linux.c  -fPIC -DPIC -o x86_64/.libs/Los-linux.o
In file included from x86_64/Los-linux.c:4:
x86_64/Gos-linux.c: In function '_ULx86_64_sigreturn':
x86_64/Gos-linux.c:141:27: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  141 |   struct sigcontext *sc = (struct sigcontext *) c->sigcontext_addr;
      |                           ^
x86_64/Gos-linux.c: Assembler messages:
x86_64/Gos-linux.c:149: Error: unsupported instruction `mov'
make[2]: *** [Makefile:3908: x86_64/Los-linux.lo] Error 1
make[2]: Leaving directory '/root/libunwind/src'
make[1]: *** [Makefile:2211: all] Error 2
make[1]: Leaving directory '/root/libunwind/src'
make: *** [Makefile:607: all-recursive] Error 1
```

x32 ABI (https://en.wikipedia.org/wiki/X32_ABI) is supported at least by `gcc` with `-mx32` or when gcc is built with ` --host=x86_64-pc-linux-gnux32` (Gentoo uses both forms).

--- a/src/x86_64/Gos-linux.c
+++ b/src/x86_64/Gos-linux.c
@@ -149,7 +149,7 @@ x86_64_sigreturn (unw_cursor_t *cursor)
   __asm__ __volatile__ ("mov %0, %%rsp;"
                         "mov %1, %%rax;"
                         "syscall"
-                        :: "r"(sc), "i"(SYS_rt_sigreturn)
+                        :: "r"((uint64_t)sc), "i"(SYS_rt_sigreturn)
                         : "memory");
   abort();
 }

